core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java

```diff
@@ -0,0 +1,254 @@
+        MethodHandle outputFunction = OUTPUT_FUNCTION.bindTo(type);
+        Class<? extends AccumulatorState> stateInterface = ListaggAggregationState.class;
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, type.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                inputParameterMetadata,
+                inputFunction,
+                Optional.empty(),
+                combineFunction,
+                outputFunction,
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        stateInterface,
+                        stateSerializer,
+                        stateFactory)),
+                outputType);
+
+        GenericAccumulatorFactoryBinder factory = AccumulatorCompiler.generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, factory);
+    }
+
+    public static void input(Type type, ListaggAggregationState state, Block value, Slice separator, boolean overflowError, Slice overflowFiller, boolean showOverflowEntryCount, int position)
+    {
+        if (state.isEmpty()) {
+            if (overflowFiller.length() > MAX_OVERFLOW_FILLER_LENGTH) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, format("Overflow filler length %d exceeds maximum length %d", overflowFiller.length(), MAX_OVERFLOW_FILLER_LENGTH));
+            }
+            // Set the parameters of the LISTAGG command within the state so that
+            // they can be used within the `output` function
+            state.setSeparator(separator);
+            state.setOverflowError(overflowError);
+            state.setOverflowFiller(overflowFiller);
+            state.setShowOverflowEntryCount(showOverflowEntryCount);
+        }
+        if (!value.isNull(position)) {
+            state.add(value, position);
+        }
+    }
+
+    public static void combine(Type type, ListaggAggregationState state, ListaggAggregationState otherState)
+    {
+        Slice previousSeparator = state.getSeparator();
+        if (previousSeparator == null) {
+            state.setSeparator(otherState.getSeparator());
+            state.setOverflowError(otherState.isOverflowError());
+            state.setOverflowFiller(otherState.getOverflowFiller());
+            state.setShowOverflowEntryCount(otherState.showOverflowEntryCount());
+        }
+
+        state.merge(otherState);
+    }
+
+    public static void output(Type type, ListaggAggregationState state, BlockBuilder out)
+    {
+        if (state.isEmpty()) {
+            out.appendNull();
+        }
+        else {
+            outputState(state, out, MAX_OUTPUT_LENGTH);
+        }
+    }
+
+    @VisibleForTesting
+    protected static void outputState(ListaggAggregationState state, BlockBuilder out, int maxOutputLength)
+    {
```

Comment:
Here's a variant that tries to keep the values+separators within MAX_OUTPUT_LENGTH but potentially exceeds it when appending the separator, filter and count. The code is very straightforward and easy to reason about:


```java
Slice separator = state.getSeparator();
int separatorLength = separator.length();
OutputContext context = new OutputContext();
state.forEach((block, position) -> {
    int entryLength = block.getSliceLength(position);
    int spaceRequired = entryLength + (context.emittedEntryCount > 0 ? separatorLength : 0);

    if (context.outputLength + spaceRequired > maxOutputLength) {
        context.overflow = true;
        return false;
    }

    if (context.emittedEntryCount > 0) {
        out.writeBytes(separator, 0, separatorLength);
        context.outputLength += separatorLength;
    }

    block.writeBytesTo(position, 0, entryLength, out);
    context.outputLength += entryLength;
    context.emittedEntryCount++;

    return true;
});

if (context.overflow) {
    if (state.isOverflowError()) {
        throw new TrinoException(EXCEEDED_FUNCTION_MEMORY_LIMIT, format("Concatenated string has the length in bytes larger than the maximum output length %d", maxOutputLength));
    }

    if (context.emittedEntryCount > 0) {
        out.writeBytes(separator, 0, separatorLength);
    }
    out.writeBytes(state.getOverflowFiller(), 0, state.getOverflowFiller().length());

    if (state.showOverflowEntryCount()) {
        out.writeBytes(Slices.utf8Slice("("), 0, 1);
        Slice count = Slices.utf8Slice(Integer.toString(state.getEntryCount() - context.emittedEntryCount));
        out.writeBytes(count, 0, count.length());
        out.writeBytes(Slices.utf8Slice(")"), 0, 1);
    }
}

out.closeEntry();
```

Reply:
Definitely. Thank you for pointing out the readability improvements.